exec/java-exec/src/main/java/org/apache/drill/exec/cache/VectorSerializer.java

```diff
@@ -62,27 +72,65 @@ public Writer write(VectorAccessible va) throws IOException {
 
     @SuppressWarnings("resource")
     public Writer write(VectorAccessible va, SelectionVector2 sv2) throws IOException {
+      checkNotNull(va);
       WritableBatch batch = WritableBatch.getBatchNoHVWrap(
           va.getRecordCount(), va, sv2 != null);
       return write(batch, sv2);
     }
 
     public Writer write(WritableBatch batch, SelectionVector2 sv2) throws IOException {
-      VectorAccessibleSerializable vas;
-      if (sv2 == null) {
-        vas = new VectorAccessibleSerializable(batch, allocator);
-      } else {
-        vas = new VectorAccessibleSerializable(batch, sv2, allocator);
-      }
-      if (retain) {
-        vas.writeToStreamAndRetain(stream);
-      } else {
-        vas.writeToStream(stream);
+      return write(batch, sv2, false);
+    }
+
+    public Writer write(WritableBatch batch, SelectionVector2 sv2, boolean retain) throws IOException {
+      checkNotNull(batch);
+      checkNotNull(channel);
+      final Timer.Context timerContext = metrics.timer(WRITER_TIMER).time();
+
+      final DrillBuf[] incomingBuffers = batch.getBuffers();
+      final UserBitShared.RecordBatchDef batchDef = batch.getDef();
+
+      try {
+        /* Write the metadata to the file */
+        batchDef.writeDelimitedTo(output);
+
+
+        /* If we have a selection vector, dump it to file first */
+        if (sv2 != null) {
+          final int dataLength = sv2.getCount() * SelectionVector2.RECORD_SIZE;
+          channel.write(sv2.getBuffer(false).nioBuffer(0, dataLength));
+        }
+
+        /* Dump the array of ByteBuf's associated with the value vectors */
+        for (DrillBuf buf : incomingBuffers) {
+          /* dump the buffer into the OutputStream */
+          channel.write(buf.nioBuffer());
+        }
+
+        timeNs += timerContext.stop();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      } finally {
+        if (!retain) {
```

Comment:
Frankly, the caller should decide what to do with the SV2 and the batch. Much simpler to just do:

```
writer.write(batch, sv2);
batch.clear();
sv2.clear();
```

Or, provide a `writeAndClear()` method if we think it is worth pulling those two lines into the write method.